quil-plotting/quil_plotting/dataframe.py

```diff
@@ -0,0 +1,346 @@
+        sorted_df = df.sort_values(["Time (s)"])
+        runner_values = sorted_df[runners].drop_duplicates()
+    elif runner_order == "Topological":
+        # need the root instructions
+        raise NotImplementedError()
+    else:
+        runner_values = df[runners].drop_duplicates()
+    runner_map = {runner: idx for idx, runner in enumerate(runner_values)}
+    df["Offset"] = df[runners].apply(lambda runner: runner_map.get(runner, 0))
+
+    # Calculate colors
+    color_cycle = cycle(Color)
+    color_values = df[color_by].cat.categories
+    # pre-defined colors
+    color_map = {"ISWAP": Color.YELLOW, "RX": Color.GREEN, "CZ": Color.BLUE}
+    # add additional colors as needed
+    color_map |= {val: next(color_cycle) for val in color_values if val not in color_map}
+    df["Color"] = df[color_by].apply(lambda x: hex_to_rgba(color_map.get(x, "#ffffff")))
+    df["Color Faded"] = df[color_by].apply(lambda x: hex_to_rgba(color_map.get(x, "#ffffff"), 0.3))
+
+    # Normalize
+    if normalize_by is None:
+        df["Normalized IQ"] = df["IQ"]
+    else:
+        df["Magnitude"] = np.abs(df["IQ"])
+        normalization_values = df.groupby(normalize_by, observed=True)["Magnitude"].transform("max")
+        df["Normalized IQ"] = 0.45 * df["IQ"] / normalization_values
+
+    # Label
+    if label_by is None:
+        label_by = runners
+
+    df["Label"] = df[label_by]
+    df["Runner"] = df[runners]
+
+    df = df.astype(
+        {
+            "Instruction Index": int,
+            "Instruction": "category",
+            "Logical Instruction": "category",
+            "Gate": "category",
+            "Qubit": "category",
+            "Pulse": "category",
+            "Waveform": "category",
+            "Start Time (s)": float,
+            "End Time (s)": float,
+            "Duration (s)": float,
+            "Pulse Time (s)": float,
+            "Frame": "category",
+            "Channel Type": "category",
+            "IQ": complex,
+            "Offset": float,
+            "Color": "category",
+            "Color Faded": "category",
+            "Normalized IQ": complex,
+            "Label": "category",
+            "Runner": "category",
+        }
+    )
+
+    return df
+
+
+def deoxidize(num) -> complex:
+    """Remove Rust from a Number or PrefixExpression."""
+    num = num.inner()
+    if isinstance(num, PrefixExpression):
+        if num.operator == PrefixOperator.Plus:
+            return num.expression.evaluate({}, {})
+        elif num.operator == PrefixOperator.Minus:
+            return -1 * num.expression.evaluate({}, {})
+        else:
+            raise ValueError()
+    return num
+
+
+def get_iqs(pulse: Pulse, waveforms: Dict, frame):
+    waveform_reference = pulse.waveform
+    name = waveform_reference.name
+    parameters = pulse.waveform.parameters
+    sample_rate = frame["SAMPLE-RATE"].inner().to_real()
+    match name:
+        case "flat":
+            waveform_envelope = FlatWaveform(**parameters)
```

Comment:
Each of the template waveforms have an `into_iq_values` method, any reason that isn't used here?